decompressor/lz4/lz4.go

```diff
@@ -0,0 +1,95 @@
+package lz4
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pierrec/lz4"
+	"io"
+)
+
+type Reader struct {
+	src  io.Reader
+	data []byte
+}
+
+func NewReader(src io.Reader) *Reader {
+	return &Reader{src: src}
+}
+
+var lz4Type = []byte{0x82}
+
+func (r *Reader) Read(buf []byte) (n int, err error) {
+	// exhaust remaining data from previous Read()
+	if len(r.data) > 0 {
+		n = copy(buf, r.data)
+		r.data = r.data[n:]
+		if len(r.data) == 0 {
+			r.data = nil
+		}
+		return
+	}
+
+	// checksum
+	b := make([]byte, 16)
+	n, err = r.src.Read(b)
+	if err != nil {
+		return 0, err
+	}
+	if n < 16 {
+		return 0, fmt.Errorf("invalid checksum read")
+	}
+	magic := make([]byte, 1)
+	_, err = r.src.Read(magic)
+	if err != nil {
+		return 0, err
+	}
+	if !bytes.Equal(magic, lz4Type) {
```

Comment:
This isn't `magic` - this is compression type :)
Currently `clickhouse` supports two compression types - `lz4` and `zstd` - see [the corresponding source code](https://github.com/yandex/ClickHouse/blob/ae8783aee3ef982b6eb7e1721dac4cb3ce73f0fe/dbms/src/IO/CompressedStream.h#L24), so it would be great if the `Reader` would support `zstd` too.